### DISCUSSION TITLE
How to properly call an async function inside state management with tauri command?

### DISCUSSION THREAD

**Discussion by Hellager** [tauri-apps/tauri#6531] [Q&A] [answered]
when i try call an async function inside state management  using tauri command like this
```
impl SomeHandler {
  pub async fn something_async {
    println!("do something!");
  }
}

#[tauri::command]
pub async fn do_something_async<R: Runtime>(state: State<'_, SomeHandlerWrapper>) {
  state.0.lock().unwrap().something_async().await;
}
```

It will failed to compile and get this error:
```
future is not `Send` as this value is used across an await

consider moving this into a `let` binding to create a shorter lived borrow
```

What's the proper way to do this?

**Comment by FabianLars** [comment, marked as answer]
One simple workaround would be to use tokio's mutex which _can_ be held across an await https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html (also re-exported as `tauri::async_runtime::Mutex`

**Reply by Hellager** [reply]
That's it! Thanks a lot.
For anyone who may have the same question, this is how it should look like
The `Result<(), ()>` is in need, you can check this discussion https://github.com/tauri-apps/tauri/discussions/4317
```rust
use tauri::async_runtime::Mutex;
impl SomeHandler {
  pub async fn something_async() -> Result<(), ()>{
    println!("do something!");
    Ok(())
  }
}

#[tauri::command]
pub async fn execute_some_command(state: State<'_, SomeHandlerWrapper>) -> Result<(), ()> {
  let mut cur_state = state.0.lock().await;
  cur_state.do_something_async().await
}
```
